The job **`dist-x86_64-msvc`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/28424959309/job/84225966673) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/84225966673) [(plain)](https://github.com/rust-lang/rust/commit/e103f503bd3c5b23cce589e8c1ce06b8d5738eb1/checks/84225966673/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling rustc_interface v0.0.0 (C:\a\rust\rust\compiler\rustc_interface)
   Compiling rustc_driver_impl v0.0.0 (C:\a\rust\rust\compiler\rustc_driver_impl)
error: linking with `link.exe` failed: exit code: 4044
  |
  = note: "C:\\Program Files\\Microsoft Visual Studio\\18\\Enterprise\\VC\\Tools\\MSVC\\14.51.36231\\bin\\HostX64\\x64\\link.exe" "/NOLOGO" "C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-rustc\\x86_64-pc-windows-msvc\\release\\build\\rustc-main/8d9eea4b9618f4d4\\out\\rustcchUNRi\\symbols.o" "<1 object files omitted>" "C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-rustc\\x86_64-pc-windows-msvc\\release\\build\\rustc_driver\\e582c7f1d7e04a53\\out\\rustc_driver-e582c7f1d7e04a53.dll.lib" "<sysroot>\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib\\std-429ad58a64234ee8.dll.lib" "<sysroot>\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib/libcompiler_builtins-*.rlib" "psapi.lib" "shell32.lib" "ole32.lib" "uuid.lib" "advapi32.lib" "ws2_32.lib" "ntdll.lib" "bcrypt.lib" "advapi32.lib" "kernel32.lib" "advapi32.lib" "ole32.lib" "oleaut32.lib" "legacy_stdio_definitions.lib" "kernel32.lib" "kernel32.lib" "kernel32.lib" "kernel32.lib" "ntdll.lib" "userenv.lib" "ws2_32.lib" "dbghelp.lib" "/defaultlib:libcmt" "C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-rustc\\x86_64-pc-windows-msvc\\release\\build\\rustc-main/8d9eea4b9618f4d4\\out\\rustcchUNRi\\advapi32.dll_imports_indirect.lib" "C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-rustc\\x86_64-pc-windows-msvc\\release\\build\\rustc-main/8d9eea4b9618f4d4\\out\\rustcchUNRi\\api-ms-win-core-errorhandling-l1-1-3.dll_imports_indirect.lib" "C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-rustc\\x86_64-pc-windows-msvc\\release\\build\\rustc-main/8d9eea4b9618f4d4\\out\\rustcchUNRi\\api-ms-win-core-file-fromapp-l1-1-0.dll_imports_indirect.lib" "C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-rustc\\x86_64-pc-windows-msvc\\release\\build\\rustc-main/8d9eea4b9618f4d4\\out\\rustcchUNRi\\api-ms-win-core-handle-l1-1-0.dll_imports_indirect.lib" "C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-rustc\\x86_64-pc-windows-msvc\\release\\build\\rustc-main/8d9eea4b9618f4d4\\out\\rustcchUNRi\\api-ms-win-core-ioring-l1-1-0.dll_imports_indirect.lib" "C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-rustc\\x86_64-pc-windows-msvc\\release\\build\\rustc-main/8d9eea4b9618f4d4\\out\\rustcchUNRi\\api-ms-win-core-libraryloader-l2-1-0.dll_imports_indirect.lib" "C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-rustc\\x86_64-pc-windows-msvc\\release\\build\\rustc-main/8d9eea4b9618f4d4\\out\\rustcchUNRi\\api-ms-win-core-memory-l1-1-3.dll_imports_indirect.lib" "C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-rustc\\x86_64-pc-windows-msvc\\release\\build\\rustc-main/8d9eea4b9618f4d4\\out\\rustcchUNRi\\api-ms-win-core-memory-l1-1-4.dll_imports_indirect.lib" "C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-rustc\\x86_64-pc-windows-msvc\\release\\build\\rustc-main/8d9eea4b9618f4d4\\out\\rustcchUNRi\\api-ms-win-core-memory-l1-1-5.dll_imports_indirect.lib" "C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-rustc\\x86_64-pc-windows-msvc\\release\\build\\rustc-main/8d9eea4b9618f4d4\\out\\rustcchUNRi\\api-ms-win-core-memory-l1-1-6.dll_imports_indirect.lib" "C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-rustc\\x86_64-pc-windows-msvc\\release\\build\\rustc-main/8d9eea4b9618f4d4\\out\\rustcchUNRi\\api-ms-win-core-memory-l1-1-7.dll_imports_indirect.lib" "C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-rustc\\x86_64-pc-windows-msvc\\release\\build\\rustc-main/8d9eea4b9618f4d4\\out\\rustcchUNRi\\api-ms-win-core-memory-l1-1-8.dll_imports_indirect.lib" "C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-rustc\\x86_64-pc-windows-msvc\\release\\build\\rustc-main/8d9eea4b9618f4d4\\out\\rustcchUNRi\\api-ms-win-core-synch-l1-2-0.dll_imports_indirect.lib" "C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-rustc\\x86_64-pc-windows-msvc\\release\\build\\rustc-main/8d9eea4b9618f4d4\\out\\rustcchUNRi\\api-ms-win-core-util-l1-1-1.dll_imports_indirect.lib" "C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-rustc\\x86_64-pc-windows-msvc\\release\\build\\rustc-main/8d9eea4b9618f4d4\\out\\rustcchUNRi\\api-ms-win-core-winrt-error-l1-1-0.dll_imports_indirect.lib" "C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-rustc\\x86_64-pc-windows-msvc\\release\\build\\rustc-main/8d9eea4b9618f4d4\\out\\rustcchUNRi\\api-ms-win-core-winrt-l1-1-0.dll_imports_indirect.lib" "C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-rustc\\x86_64-pc-windows-msvc\\release\\build\\rustc-main/8d9eea4b9618f4d4\\out\\rustcchUNRi\\api-ms-win-core-wow64-l1-1-1.dll_imports_indirect.lib" "C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-rustc\\x86_64-pc-windows-msvc\\release\\build\\rustc-main/8d9eea4b9618f4d4\\out\\rustcchUNRi\\api-ms-win-security-base-l1-2-2.dll_imports_indirect.lib" "C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-rustc\\x86_64-pc-windows-msvc\\release\\build\\rustc-main/8d9eea4b9618f4d4\\out\\rustcchUNRi\\avrt.dll_imports_indirect.lib" "C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-rustc\\x86_64-pc-windows-msvc\\release\\build\\rustc-main/8d9eea4b9618f4d4\\out\\rustcchUNRi\\bcp47mrm.dll_imports_indirect.lib" "C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-rustc\\x86_64-pc-windows-msvc\\release\\build\\rustc-main/8d9eea4b9618f4d4\\out\\rustcchUNRi\\bcryptprimitives.dll_imports_indirect.lib" "C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-rustc\\x86_64-pc-windows-msvc\\release\\build\\rustc-main/8d9eea4b9618f4d4\\out\\rustcchUNRi\\clfsw32.dll_imports_indirect.lib" "C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-rustc\\x86_64-pc-windows-msvc\\release\\build\\rustc-main/8d9eea4b9618f4d4\\out\\rustcchUNRi\\dbghelp.dll_imports_indirect.lib" "C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-rustc\\x86_64-pc-windows-msvc\\release\\build\\rustc-main/8d9eea4b9618f4d4\\out\\rustcchUNRi\\elscore.dll_imports_indirect.lib" "C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-rustc\\x86_64-pc-windows-msvc\\release\\build\\rustc-main/8d9eea4b9618f4d4\\out\\rustcchUNRi\\gdi32.dll_imports_indirect.lib" "C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-rustc\\x86_64-pc-windows-msvc\\release\\build\\rustc-main/8d9eea4b9618f4d4\\out\\rustcchUNRi\\icu.dll_imports_indirect.lib" "C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-rustc\\x86_64-pc-windows-msvc\\release\\build\\rustc-main/8d9eea4b9618f4d4\\out\\rustcchUNRi\\icuin.dll_imports_indirect.lib" "C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-rustc\\x86_64-pc-windows-msvc\\release\\build\\rustc-main/8d9eea4b9618f4d4\\out\\rustcchUNRi\\icuuc.dll_imports_indirect.lib" "C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-rustc\\x86_64-pc-windows-msvc\\release\\build\\rustc-main/8d9eea4b9618f4d4\\out\\rustcchUNRi\\imagehlp.dll_imports_indirect.lib" "C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-rustc\\x86_64-pc-windows-msvc\\release\\build\\rustc-main/8d9eea4b9618f4d4\\out\\rustcchUNRi\\kernel32.dll_imports_indirect.lib" "C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-rustc\\x86_64-pc-windows-msvc\\release\\build\\rustc-main/8d9eea4b9618f4d4\\out\\rustcchUNRi\\ktmw32.dll_imports_indirect.lib" "C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-rustc\\x86_64-pc-windows-msvc\\release\\build\\rustc-main/8d9eea4b9618f4d4\\out\\rustcchUNRi\\netapi32.dll_imports_indirect.lib" "C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-rustc\\x86_64-pc-windows-msvc\\release\\build\\rustc-main/8d9eea4b9618f4d4\\out\\rustcchUNRi\\normaliz.dll_imports_indirect.lib" "C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-rustc\\x86_64-pc-windows-msvc\\release\\build\\rustc-main/8d9eea4b9618f4d4\\out\\rustcchUNRi\\ntdll.dll_imports_indirect.lib" "C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-rustc\\x86_64-pc-windows-msvc\\release\\build\\rustc-main/8d9eea4b9618f4d4\\out\\rustcchUNRi\\ole32.dll_imports_indirect.lib" "C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-rustc\\x86_64-pc-windows-msvc\\release\\build\\rustc-main/8d9eea4b9618f4d4\\out\\rustcchUNRi\\oleacc.dll_imports_indirect.lib" "C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-rustc\\x86_64-pc-windows-msvc\\release\\build\\rustc-main/8d9eea4b9618f4d4\\out\\rustcchUNRi\\oleaut32.dll_imports_indirect.lib" "C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-rustc\\x86_64-pc-windows-msvc\\release\\build\\rustc-main/8d9eea4b9618f4d4\\out\\rustcchUNRi\\psapi.dll_imports_indirect.lib" "C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-rustc\\x86_64-pc-windows-msvc\\release\\build\\rustc-main/8d9eea4b9618f4d4\\out\\rustcchUNRi\\rtworkq.dll_imports_indirect.lib" "C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-rustc\\x86_64-pc-windows-msvc\\release\\build\\rustc-main/8d9eea4b9618f4d4\\out\\rustcchUNRi\\txfw32.dll_imports_indirect.lib" "C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-rustc\\x86_64-pc-windows-msvc\\release\\build\\rustc-main/8d9eea4b9618f4d4\\out\\rustcchUNRi\\user32.dll_imports_indirect.lib" "C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-rustc\\x86_64-pc-windows-msvc\\release\\build\\rustc-main/8d9eea4b9618f4d4\\out\\rustcchUNRi\\usp10.dll_imports_indirect.lib" "C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-rustc\\x86_64-pc-windows-msvc\\release\\build\\rustc-main/8d9eea4b9618f4d4\\out\\rustcchUNRi\\version.dll_imports_indirect.lib" "C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-rustc\\x86_64-pc-windows-msvc\\release\\build\\rustc-main/8d9eea4b9618f4d4\\out\\rustcchUNRi\\wofutil.dll_imports_indirect.lib" "/NXCOMPAT" "/LIBPATH:C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-rustc\\x86_64-pc-windows-msvc\\release\\build\\stacker/0e91dbd10266030c\\out" "/LIBPATH:C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-rustc\\x86_64-pc-windows-msvc\\release\\build\\psm/a1024b6746e4986e\\out" "/LIBPATH:C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-rustc\\x86_64-pc-windows-msvc\\release\\build\\blake3/31d2e063d249363b\\out" "/LIBPATH:C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-rustc\\x86_64-pc-windows-msvc\\release\\build\\rustc_llvm/7fbd1cb716513acd\\out" "/LIBPATH:C:\\Program Files\\Microsoft Visual Studio\\18\\Enterprise\\VC\\Tools\\MSVC\\14.51.36231\\atlmfc\\lib\\x64" "/LIBPATH:C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\llvm\\lib" "/OUT:C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-rustc\\x86_64-pc-windows-msvc\\release\\build\\rustc-main/8d9eea4b9618f4d4\\out\\rustc_main-8d9eea4b9618f4d4.exe" "/OPT:REF,ICF" "/DEBUG" "/PDBALTPATH:%_PDB%" "/NATVIS:<sysroot>\\lib\\rustlib\\etc\\intrinsic.natvis" "/NATVIS:<sysroot>\\lib\\rustlib\\etc\\liballoc.natvis" "/NATVIS:<sysroot>\\lib\\rustlib\\etc\\libcore.natvis" "/NATVIS:<sysroot>\\lib\\rustlib\\etc\\libstd.natvis" "/NATVIS:C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-rustc\\x86_64-pc-windows-msvc\\release\\build\\rustc-main/8d9eea4b9618f4d4\\out\\rustcchUNRi\\rustc_main-0.natvis" "/NATVIS:C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-rustc\\x86_64-pc-windows-msvc\\release\\build\\rustc-main/8d9eea4b9618f4d4\\out\\rustcchUNRi\\rustc_main-1.natvis" "/NATVIS:C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-rustc\\x86_64-pc-windows-msvc\\release\\build\\rustc-main/8d9eea4b9618f4d4\\out\\rustcchUNRi\\rustc_main-2.natvis" "-Wl,--compress-debug-sections=zlib" "C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\stage1-rustc\\x86_64-pc-windows-msvc\\release\\build\\rustc-main/5a5955a99bc88b21\\out\\resources\\rustc_main_resource.res" "/MANIFEST:EMBED" "/MANIFESTINPUT:C:\\a\\rust\\rust\\compiler\\rustc\\Windows Manifest.xml" "/WX"
  = note: some arguments are omitted. use `--verbose` to show all linker arguments
  = note: LINK : warning LNK4044: unrecognized option '/Wl,--compress-debug-sections=zlib'; ignored
          LINK : error LNK1218: warning treated as error; no output file generated
          

error: could not compile `rustc-main` (bin "rustc-main") due to 1 previous error
Bootstrap failed while executing `build --target x86_64-pc-windows-msvc --host x86_64-pc-windows-msvc --stage 2 library/std --rust-profile-generate C:\a\rust\rust\opt-artifacts\rustc-pgo`
Build completed unsuccessfully in 0:43:29
```

</details>

--- a/bootstrap.example.toml
+++ b/bootstrap.example.toml
@@ -698,6 +698,7 @@
 #rust.debuginfo-level-tests = 0
 
 # Compress debuginfo of Rust and C/C++ code.
+# Currently, this only works on Linux.
 # Valid options:
 # - "off" or false: disable compression
 # - true: compress debuginfo with the default compression method (currently zlib)
@@ -1027,6 +1028,7 @@
 #split-debuginfo = if linux || windows-gnu { off } else if windows-msvc { packed } else if apple { unpacked }
 
 # Compress debuginfo for Rust and C/C++ code of this target.
+# Currently, this only works on Linux.
 # Valid options:
 # - "off" or false: disable compression
 # - true: compress debuginfo with the default compression method (currently zlib)

--- a/src/bootstrap/src/core/builder/cargo.rs
+++ b/src/bootstrap/src/core/builder/cargo.rs
@@ -340,18 +340,23 @@ impl Cargo {
             self.rustdocflags.arg(&arg);
         }
 
-        if !builder.config.dry_run() {
-            match builder.config.compress_debuginfo(target) {
-                CompressDebuginfo::Zlib => {
-                    // Do not enable Zlib compression on:
-                    // - Windows, because MSVC/PDB doesn't support it
-                    // - macOS, because its linker doesn't know the flag
-                    if !self.target.is_windows() && !self.target.is_apple() {
+        match builder.config.compress_debuginfo(target) {
+            CompressDebuginfo::Zlib => {
+                // Do not enable Zlib compression on:
+                // - Windows, because MSVC/PDB doesn't support it
+                // - macOS, because its linker doesn't know the flag
+                if !self.target.is_windows() && !self.target.is_apple() {
+                    // If we link through cc, we need the -Wl prefix.
+                    // If we don't, then we must not add it, because the linker wouldn't
+                    // understand it.
+                    if helpers::use_host_linker(target) {
                         self.rustflags.arg("-Clink-arg=-Wl,--compress-debug-sections=zlib");
+                    } else {
+                        self.rustflags.arg("-Clink-arg=--compress-debug-sections=zlib");
                     }
                 }
-                CompressDebuginfo::Off => {}
             }
+            CompressDebuginfo::Off => {}
         }
 
         // Ignore linker warnings for now. These are complicated to fix and don't affect the build.

--- a/src/bootstrap/src/core/builder/cargo.rs
+++ b/src/bootstrap/src/core/builder/cargo.rs
@@ -343,7 +343,12 @@ impl Cargo {
         if !builder.config.dry_run() {
             match builder.config.compress_debuginfo(target) {
                 CompressDebuginfo::Zlib => {
-                    self.rustflags.arg("-Clink-arg=-Wl,--compress-debug-sections=zlib");
+                    // Do not enable Zlib compression on:
+                    // - Windows, because MSVC/PDB doesn't support it
+                    // - macOS, because its linker doesn't know the flag
+                    if !self.target.is_windows() && !self.target.is_apple() {
+                        self.rustflags.arg("-Clink-arg=-Wl,--compress-debug-sections=zlib");
+                    }
                 }
                 CompressDebuginfo::Off => {}
             }

--- a/src/bootstrap/src/core/config/target_selection.rs
+++ b/src/bootstrap/src/core/config/target_selection.rs
@@ -82,6 +82,10 @@ impl TargetSelection {
         self.contains("windows")
     }
 
+    pub fn is_apple(&self) -> bool {
+        self.contains("apple")
+    }
+
     pub fn is_windows_gnu(&self) -> bool {
         self.ends_with("windows-gnu")
     }

